core-processor/src/main/java/io/micronaut/expressions/parser/ast/access/ElementMethodCall.java

```diff
@@ -0,0 +1,120 @@
+import io.micronaut.inject.ast.ClassElement;
+import io.micronaut.inject.ast.ElementQuery;
+import io.micronaut.inject.ast.MethodElement;
+import io.micronaut.inject.visitor.VisitorContext;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.commons.GeneratorAdapter;
+import org.objectweb.asm.commons.Method;
+
+import java.util.List;
+
+import static io.micronaut.expressions.parser.ast.util.EvaluatedExpressionCompilationUtils.getRequiredClassElement;
+import static org.objectweb.asm.Opcodes.INVOKESTATIC;
+
+/**
+ * Expression AST node used for method invocation.
+ * This node represents both object method invocation and static method
+ * invocation
+ *
+ * @author Sergey Gavrilov
+ * @since 4.0.0
+ */
+@Internal
+public class ElementMethodCall extends AbstractMethodCall {
+    protected final ExpressionNode callee;
+
+    public ElementMethodCall(ExpressionNode callee,
+                             String name,
+                             List<ExpressionNode> arguments) {
+        super(name, arguments);
+        this.callee = callee;
+    }
+
+    @Override
+    protected void generateBytecode(ExpressionCompilationContext ctx) {
+        GeneratorAdapter mv = ctx.methodVisitor();
+        VisitorContext visitorContext = ctx.visitorContext();
+        Type calleeType = callee.resolveType(ctx);
+        Method method = usedMethod.toAsmMethod();
+
+        ClassElement calleeClass = getRequiredClassElement(calleeType, visitorContext);
+
+        if (callee instanceof TypeIdentifier) {
+            compileArguments(ctx);
+            if (calleeClass.isInterface()) {
+                mv.visitMethodInsn(INVOKESTATIC, calleeType.getInternalName(), name,
+                    usedMethod.getDescriptor(), true);
+            } else {
+                mv.invokeStatic(calleeType, method);
+            }
+        } else {
+            callee.compile(ctx);
+            compileArguments(ctx);
+            if (calleeClass.isInterface()) {
+                mv.invokeInterface(calleeType, method);
+            } else {
+                mv.invokeVirtual(calleeType, method);
+            }
+        }
+    }
+
+    @Override
+    protected CandidateMethod resolveUsedMethod(ExpressionCompilationContext ctx) {
+        List<Type> argumentTypes = resolveArgumentTypes(ctx);
+        Type calleeType = callee.resolveType(ctx);
+
+        ElementQuery<MethodElement> methodQuery = buildMethodQuery();
+        List<CandidateMethod> candidateMethods =
+            ctx.visitorContext()
+                .getClassElement(calleeType.getClassName())
+                .stream()
+                .flatMap(element -> element.getEnclosedElements(methodQuery).stream())
+                .map(method -> toCandidateMethod(ctx, method, argumentTypes))
+                .filter(method -> method.isMatching(ctx.visitorContext()))
+                .toList();
+
+        if (candidateMethods.isEmpty()) {
+            throw new ExpressionCompilationException(
```

Comment:
when throwing these exceptions you may need to pass the originating element so where they are caught you can report an appropriate error that points to the source code

Reply:
All parsing and compilation exceptions are caught in single point in `EvaluatedExpressionWriter` class which has access to originating element, so when we fail compilation using `VisitorContext#fail`, that element is passed